src/SqlProviderBase.ts

```diff
@@ -398,20 +401,28 @@ export abstract class SqlProviderBase extends NoSqlProvider.DbProvider {
                                     return trans.runQuery('ALTER TABLE ' + storeSchema.name + ' RENAME TO temp_' + storeSchema.name);
                                 });
 
-                                // Migrate the data over using our existing put functions
-                                // (since it will do the right things with the indexes)
-                                // and delete the temp table.
-                                let migrator = () => {
+                                // Migrate the data over using our existing put functions (since it will do the right things with 
+                                // the indexes) and delete the temp table. Run the upgrade in batches to prevent OOM issues
+                                let migrator = (batchOffset = 0): SyncTasks.Promise<any> => {
+                                    let esimatedSize = storeSchema.estimatedObjBytes || DB_SIZE_ESIMATE_DEFAULT;
+                                    let batchSize = Math.max(1, Math.floor(DB_MIGRATION_MAX_BYTE_TARGET / esimatedSize));
                                     let store = trans.getStore(storeSchema.name);
-                                    return trans.internal_getResultsFromQuery('SELECT nsp_data FROM temp_' + storeSchema.name)
-                                    .then(objs => {
-                                        return store.put(objs).then(() => {
-                                            return trans.runQuery('DROP TABLE temp_' + storeSchema.name);
-                                        });
+                                    return trans.internal_getResultsFromQuery('SELECT nsp_data FROM temp_' + storeSchema.name + ' LIMIT ' +
+                                            batchSize + ' OFFSET ' + batchOffset)
+                                        .then(objs => {
+                                            return store.put(objs).then(() => {
+                                                // Are we done migrating?
+                                                if (objs.length < batchSize) {
+                                                    return trans.runQuery('DROP TABLE temp_' + storeSchema.name);
+                                                }
+                                                return migrator(batchOffset + batchSize);
```

Comment:
If we happened to get undefined out of the db for a key, it's possible that it would be _.compact'd away and give us an off-by-one error. I'd rather rely on SQL truthfullness here